Classes/MapEditor/Stats/MapStats.cs

```diff
@@ -0,0 +1,125 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Edda.Classes.MapEditor.Stats {
+    public class MapStats {
+        /// <summary>
+        /// Song properties needed for calculations
+        /// </summary>
+        public double globalBPM;
+        public double songDuration;
+
+        /// <summary>
+        /// Notes
+        /// </summary>
+        public int allNotes;
+        public int selectedNotes;
+        public int singleNotes;
+        public int doubleNotes;
+
+        /// <summary>
+        /// Column variety
+        /// </summary>
+        public int[] columnCounts = new int[4];
+        public double[] columnPercentages = new double[4];
+
+        /// <summary>
+        /// NPS
+        /// </summary>
+        public double npsSong;
+        public double npsMapped;
+        public double nps16Beat;
+        public double nps8Beat;
+        public double nps4Beat;
+
+        public MapStats(double globalBPM, double songDuration) {
+            this.globalBPM = globalBPM;
+            this.songDuration = songDuration;
+        }
+
+        public void Recalculate(SortedSet<Note> notes, SortedSet<Note> selectedNotes) {
+            RecalculateNotes(notes, selectedNotes);
+            RecalculateColumnVariety(notes);
+            RecalculateNPS(notes);
+        }
+
+        public void RecalculateNotes(SortedSet<Note> notes, SortedSet<Note> selectedNotes) {
+            this.allNotes = notes.Count;
+            this.selectedNotes = selectedNotes.Count;
+            this.singleNotes = CalculateRowNotes(notes, 1);
+            this.doubleNotes = CalculateRowNotes(notes, 2);
+        }
+
+        private int CalculateRowNotes(SortedSet<Note> notes, int rowCount) {
+            int rowNoteCount = 0;
+            var noteIt = notes.GetEnumerator();
+            Note? lineStartNote = null;
+            var lineNoteCount = 0;
+
+
+            while (noteIt.MoveNext()) {
+                if (lineStartNote == null || !Helper.DoubleApproxEqual(noteIt.Current.beat, lineStartNote.beat)) {
+                    if (lineNoteCount == rowCount) {
+                        rowNoteCount++;
+                    }
+                    lineStartNote = noteIt.Current;
+                    lineNoteCount = 1;
+                } else {
+                    lineNoteCount++;
+                }
+            }
+            // last row to consider after all the notes were processed
+            if (lineNoteCount == rowCount) {
+                rowNoteCount++;
+            }
+
+            return rowNoteCount;
```

Comment:
maybe just 

```csharp
return notes.GroupBy(note => note.beat)
    .Select(group => group.Count())
    .Count(count => count == rowCount);
```
should be sufficient

Reply:
Needed some changes to make sure floating-point rounding didn't throw off the count, but it works 👍